autocertkit/utils.py

```diff
@@ -1914,54 +1908,65 @@ def __call__(self, *args):
             signal.signal(signal.SIGALRM, old)
         return result
 
-def xml_to_dicts(xml, tag):
-    doc = minidom.parseString(xml.strip())
-    try:
-        result = []
-        elems = doc.getElementsByTagName(tag)
-        for el in elems:
-            # Note that we have to convert this dictionary to non-unicode
-            # strings, because we're being casual elsewhere.  That's why we're
-            # not just returning dict(el.attributes.items()).
-            item = {}
-            for k, v in el.attributes.items():
-                item[str(k)] = str(v)
-            result.append(item)
-    finally:
-        doc.unlink()
-    log.debug(result)
-    return result
+def json_loads(json_data):
+    def process_dict_keys(d):
+        new_d = {}
+        for key in d.iterkeys():
+            new_key = str(key.replace(" ","_"))
+            new_d[new_key] = d[key]
+        return new_d
+
+    def process_values(item):
+        if isinstance(item, unicode):
+            item = str(item)
+        elif isinstance(item, list):
+            for elem in item:
+                elem = process_values(elem)
+        elif isinstance(item, dict):
+            item = process_dict_keys(item)
+            for key in item.iterkeys():
+                item[key] = str(item[key])
+        return item
+
+    data = json.loads(json_data, object_hook=process_values)
+    return [data] if isinstance(data, dict) else data
 
 def call_ack_plugin(session, method, args={}, host=None):
     if not host:
         host = get_pool_master(session)
     log.debug("About to call plugin '%s' on host '%s' with args '%s'" % \
                 (method, host, args))
-                
-    return session.xenapi.host.call_plugin(host,
+
+    res = session.xenapi.host.call_plugin(host,
                                            'autocertkit',
                                            method,
                                            args)
+    try:
+        res = json_loads(res)
+    except ValueError:
```

Comment:
Hi Jun,

1. I am not sure how ensure_ascii=False in json.dumps will help us.
2. in statement  `jsonstring = json.dumps(unicode_data, ensure_ascii=False).encode('utf8')`, we will still get jsonstring as string and it will passed from plugin to kit as string.
I dont see any benefit here either.

unicode is being introduced when json.loads is called with plugin output string and parsing back to object.
for example:

```
>>> a="hi"
>>> b=json.dumps(a)
>>> b
'"hi"'
>>> json.loads(b)
u'hi'
```